src/metabase/driver/druid.clj

```diff
@@ -1,13 +1,213 @@
+                                   (recur true more))))
+    (catch Throwable _
+      false)))
+
+(defn- test:json-special-type
+  "Mark FIELD as `:json` if it's textual, doesn't already have a special type, the majority of it's values are non-nil, and all of its non-nil values
+   are valid serialized JSON dictionaries or arrays."
+  [driver field field-stats]
+  (if-not (and (not (:special_type field))
+               (contains? #{:CharField :TextField} (:base_type field)))
+    ;; this field isn't suited for this test
+    field-stats
+    ;; check for json values
+    (if-not (values-are-valid-json? (take driver/max-sync-lazy-seq-results (driver/field-values-lazy-seq driver field)))
+      field-stats
+      (do
+        (log/debug (u/format-color 'green "Field '%s' looks like it contains valid JSON objects. Setting special_type to :json." (field/qualified-name field)))
+        (assoc field-stats :special-type :json, :preview-display false)))))
+
+
+(defn- test:new-field
+  "Do the various tests that should only be done for a new `Field`.
+   We only run most of the field analysis work when the field is NEW in order to favor performance of the sync process."
+  [driver field field-stats]
+  (->> field-stats
+       (test:no-preview-display driver field)
+       (test:url-special-type   driver field)
+       (test:json-special-type  driver field)))
+
+(defn test:cardinality-and-extract-field-values-druid
+  "Extract field-values for FIELD.  If number of values exceeds `low-cardinality-threshold` then we return an empty set of values."
+  [field field-stats]
+  (println "testing for cardinality" (select-keys field [:special_type :name]))
+  ;; TODO: we need some way of marking a field as not allowing field-values so that we can skip this work if it's not appropriate
+  ;;       for example, :category fields with more than MAX values don't need to be rescanned all the time
+  (let [non-nil-values  (filter identity (druid-field-distinct-values field (inc low-cardinality-threshold)))
+        ;; only return the list if we didn't exceed our MAX values and if the the total character count of our values is reasable (#2332)
+        distinct-values (when-not (or (< low-cardinality-threshold (count non-nil-values))
+                                      ;; very simple check to see if total length of field-values exceeds (total values * max per value)
+                                      (< (* low-cardinality-threshold
+                                            field-values-entry-max-length) (reduce + (map (comp count str) non-nil-values))))
+                          non-nil-values)]
+    ;; TODO: eventually we can check for :nullable? based on the original values above
+    (cond-> (assoc field-stats :values distinct-values)
+      (and (nil? (:special_type field))
+           (pos? (count distinct-values))) (assoc :special-type :category))))
+
+(defn make-analyze-table-druid
+  "Make a generic implementation of `analyze-table`."
```

Comment:
It really doesn't make sense to have a function for creating custom `analyze-table` implementations for Druid. The function you copied & pasted is used by several other drivers to make their implementations. 

Since the Druid driver is the only driver that would be using this, the function should just be defined directly.